builder/builder-next/adapters/containerimage/pull.go

```diff
@@ -860,3 +872,41 @@ func platformMatches(img *image.Image, p *ocispec.Platform) bool {
 		Variant:      img.Variant,
 	})
 }
+
+func applySourcePolicies(ctx context.Context, str string, spls []*spb.Policy) (string, error) {
+	ref, err := cdreference.Parse(str)
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+	op := &pb.Op{
+		Op: &pb.Op_Source{
+			Source: &pb.SourceOp{
+				Identifier: srctypes.DockerImageScheme + "://" + ref.String(),
+			},
+		},
+	}
+
+	mut, err := sourcepolicy.NewEngine(spls).Evaluate(ctx, op)
+	if err != nil {
+		return "", errors.Wrap(err, "could not resolve image due to policy")
+	}
+
+	if mut {
+		var (
+			t  string
+			ok bool
+		)
+		t, newRef, ok := strings.Cut(op.GetSource().GetIdentifier(), "://")
+		if !ok {
+			return "", errors.Errorf("could not parse ref: %s", op.GetSource().GetIdentifier())
+		}
+		if ok && t != srctypes.DockerImageScheme {
+			return "", &imageutil.ResolveToNonImageError{Ref: str, Updated: newRef}
```

Comment:
This is probably fine as is since it is what is still in buildkit, but this was meant to be  the full identifier.

ref: https://github.com/moby/buildkit/pull/4209

Reply:
Ideally we wouldn't be reinventing this since it's sufficiently abstracted from the snapshotters/graphdrivers, but the implementation in BuildKit isn't really shaped right for us to re-use. I've asked @tonistiigi and he's interested in cleaning this up so we can share it; but assuming that doesn't happen immediately, would you be willing to port your fix once it's accepted in BuildKit?